### PR TITLE
Created env var for forcing node12 actions to run on node16

### LINKED ISSUE #1911: Create an env variable to force `node12` actions to run on node16
We should have a way for a self hosted runner admin to force `node12` actions to run on `node16`

```
ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION=16
```

While this could be a breaking change, it's very unlikely to impact most actions. The goal is to allow runner admins more control over what node version is used, particularly in scenarios where a node version is out of date or deprecated.

### PR DESCRIPTION
**Pull request by ruvceskistefan**
Closes #1911

### CODE REVIEW

File: src/Runner.Worker/Handlers/StepHost.cs [hunk 1]
```diff
@@ -128,7 +128,14 @@ public async Task<string> DetermineNodeRuntimeVersion(IExecutionContext executio
 
             var output = new List<string>();
             var execExitCode = await dockerManager.DockerExec(executionContext, Container.ContainerId, string.Empty, osReleaseIdCmd, output);
-            string nodeExternal;
+            
+            string nodeExternal = preferredVersion;
+            if (executionContext.Global.EnvironmentVariables != null)
```
Comment (TingluoHuang) [inline src/Runner.Worker/Handlers/StepHost.cs:133]: should we do it here or in ActionsManger.cs where we load the action.yaml?

Comment (thboop) [inline src/Runner.Worker/Handlers/StepHost.cs:133]: We should get variables from the runner machine (host context), here, not the[ job level envs](https://docs.github.com/en/actions/learn-github-actions/environment-variables)

Comment (thboop) [reply src/Runner.Worker/Handlers/StepHost.cs:133]: I think we can do something like

```
bool.TryParse(Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedActionsNodeVersion), out forceNode16);
```

Comment (thboop) [reply src/Runner.Worker/Handlers/StepHost.cs:133]: To test this, what you can do is create an `.env` file at the root of the runner layout folder, and set this value, then validate you see it works as expected

File: src/Runner.Worker/ActionManager.cs [hunk 2]
```diff
@@ -399,6 +399,13 @@ public Definition LoadAction(IExecutionContext executionContext, Pipelines.Actio
                     else if (definition.Data.Execution.ExecutionType == ActionExecutionType.NodeJS)
                     {
                         var nodeAction = definition.Data.Execution as NodeJSActionExecutionData;
+                        
+                        string nodeVersionOut = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedActionsNodeVersion);
```
Comment (TingluoHuang) [inline src/Runner.Worker/ActionManager.cs:403]: sorry... my bad...
i was suggesting `NodeScriptActionHandler` which we already have some code to force to use node16 for osx-arm64.

File: src/Runner.Common/Constants.cs [hunk 3]
```diff
@@ -238,6 +238,7 @@ public static class Agent
 
                 // Set this env var to "node12" to downgrade the node version for internal functions (e.g hashfiles). This does NOT affect the version of node actions.
                 public static readonly string ForcedInternalNodeVersion = "ACTIONS_RUNNER_FORCED_INTERNAL_NODE_VERSION"; 
+                public static readonly string ForcedActionsNodeVersion = "ACTIONS_RUNNER_FORCE_ACTIONS_NODE_VERSION";
```
Comment (TingluoHuang) [inline src/Runner.Common/Constants.cs:241]: If we suggest customer to delete `node12` from the runner, then this is not only for Actions, we might need to update the `NodeUtil.GetInternalNodeVersion` to also respect this.

might need to consider the collision case with `ACTIONS_RUNNER_FORCED_INTERNAL_NODE_VERSION` set to `node12`

File: src/Runner.Common/Util/NodeUtil.cs [hunk 4]
```diff
@@ -15,8 +15,14 @@ public static class NodeUtil
 
         public static string GetInternalNodeVersion()
         {
-            var forcedNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedInternalNodeVersion);
-            return !string.IsNullOrEmpty(forcedNodeVersion) && BuiltInNodeVersions.Contains(forcedNodeVersion) ? forcedNodeVersion : _defaultNodeVersion;
+            var forcedInternalNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedInternalNodeVersion);
+            var forcedNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedActionsNodeVersion);
+
+            return !string.IsNullOrEmpty(forcedNodeVersion) && BuiltInNodeVersions.Contains(forcedNodeVersion)
```
Comment (fhammerl) [inline src/Runner.Common/Util/NodeUtil.cs:21]: NIT: Can you extract some of these conditions into variables so the return statement is a bit easier to understand?

File: src/Runner.Worker/ActionManager.cs [hunk 5]
```diff
@@ -399,6 +399,7 @@ public Definition LoadAction(IExecutionContext executionContext, Pipelines.Actio
                     else if (definition.Data.Execution.ExecutionType == ActionExecutionType.NodeJS)
                     {
                         var nodeAction = definition.Data.Execution as NodeJSActionExecutionData;
+
```
Comment (fhammerl) [inline src/Runner.Worker/ActionManager.cs:402]:
```suggestion
```

File: src/Runner.Common/Util/NodeUtil.cs [hunk 6]
```diff
@@ -15,8 +15,21 @@ public static class NodeUtil
 
         public static string GetInternalNodeVersion()
         {
-            var forcedNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedInternalNodeVersion);
-            return !string.IsNullOrEmpty(forcedNodeVersion) && BuiltInNodeVersions.Contains(forcedNodeVersion) ? forcedNodeVersion : _defaultNodeVersion;
+            var forcedInternalNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedInternalNodeVersion);
+            var forcedNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedActionsNodeVersion);
```
Comment (fhammerl) [inline src/Runner.Common/Util/NodeUtil.cs:19]: This method is for determining the node version for **internal** JS processes, like hashfiles.js or hook execution. This has no effect on how a nodeJS github action is executed.

Why do you consider the forced **action** node version here?

Comment (ruvceskistefan) [reply src/Runner.Common/Util/NodeUtil.cs:19]: I think that you're right, now it makes more sense. I assume I misunderstood Thomas' request, thank you for your explanation.

File: src/Runner.Worker/Handlers/NodeScriptActionHandler.cs [hunk 7]
```diff
@@ -102,6 +102,11 @@ public async Task RunAsync(ActionRunStage stage)
                 Data.NodeVersion = "node16";
             }
 #endif
+            string nodeVersionOut = System.Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedActionsNodeVersion);
+            if(nodeVersionOut == "16" && Data.NodeVersion != "node16")
```
[the remark this answers is not in the document]
Comment (ruvceskistefan) [reply src/Runner.Worker/Handlers/NodeScriptActionHandler.cs:106]: Makes sense, I agree with you on this one

File: src/Runner.Common/Util/NodeUtil.cs [hunk 8]
```diff
@@ -13,10 +13,21 @@ public static class NodeUtil
         public static readonly ReadOnlyCollection<string> BuiltInNodeVersions = new(new[] { "node12", "node16" });
 #endif
 
+        public static string GetNodeVersion()
+        {
+            return System.Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedActionsNodeVersion);
```
Comment (TingluoHuang) [inline src/Runner.Common/Util/NodeUtil.cs:18]: do we need a method that only has a single line?

File: src/Test/L0/Worker/StepHostL0.cs [hunk 9]
```diff
@@ -55,7 +55,7 @@ public async Task DetermineNodeRuntimeVersionInContainerAsync()
                 Assert.Equal("node12", nodeVersion);
             }
         }
-
+
```
Comment (fhammerl) [inline src/Test/L0/Worker/StepHostL0.cs:58]: It looks like the whitespace is still there

File: src/Runner.Common/Util/NodeUtil.cs (hunk 8 above)
Comment (fhammerl) [reply src/Runner.Common/Util/NodeUtil.cs:18]: NIT: Let's remove the method and do it inline

File: src/Runner.Worker/Handlers/NodeScriptActionHandler.cs [hunk 10]
```diff
@@ -102,6 +103,11 @@ public async Task RunAsync(ActionRunStage stage)
                 Data.NodeVersion = "node16";
             }
 #endif
+            string nodeVersionOut = NodeUtil.GetNodeVersion();
```
Comment (fhammerl) [inline src/Runner.Worker/Handlers/NodeScriptActionHandler.cs:106]:
```suggestion
            var forcedNodeVersion = System.Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedActionsNodeVersion);
```

Let's remove the method above above

File: src/Runner.Common/Util/NodeUtil.cs [hunk 11]
```diff
@@ -15,8 +15,14 @@ public static class NodeUtil
 
         public static string GetInternalNodeVersion()
         {
-            var forcedNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedInternalNodeVersion);
-            return !string.IsNullOrEmpty(forcedNodeVersion) && BuiltInNodeVersions.Contains(forcedNodeVersion) ? forcedNodeVersion : _defaultNodeVersion;
+            var forcedInternalNodeVersion = Environment.GetEnvironmentVariable(Constants.Variables.Agent.ForcedInternalNodeVersion);
```
Comment (thboop) [inline src/Runner.Common/Util/NodeUtil.cs:18]: Is this just a refactor? It doesn't appear to do anything if i'm reading it correctly. Should we just revert these changes?

Comment (ruvceskistefan) [reply src/Runner.Common/Util/NodeUtil.cs:18]: @fhammerl asked to refactor this one with meaningful var names in [this](https://github.com/actions/runner/pull/1913#discussion_r886521138) comment

**Review by thboop — DISMISSED** [review]
lgtm

**Review by thboop — APPROVED** [review]
LGTM
